packages/typespec-client-generator-core/src/types.ts

```diff
@@ -1060,45 +1070,102 @@ export function getSdkModelPropertyTypeBase(
   });
 }
 
-export function getSdkModelPropertyType(
+function updateMultiPartInfo(
   context: TCGCContext,
   type: ModelProperty,
-  operation?: Operation
-): [SdkModelPropertyType, readonly Diagnostic[]] {
+  base: SdkBodyModelPropertyType,
+  operation?: Operation,
+  httpOperationPart?: HttpOperationPart
+): [void, readonly Diagnostic[]] {
   const diagnostics = createDiagnosticCollector();
-  const base = diagnostics.pipe(getSdkModelPropertyTypeBase(context, type, operation));
+  const isBytesInput =
+    base.type.kind === "bytes" ||
+    (base.type.kind === "array" && base.type.valueType.kind === "bytes");
 
-  if (isSdkHttpParameter(context, type)) return getSdkHttpParameter(context, type, operation!);
-  // I'm a body model property
-  let operationIsMultipart = false;
-  if (operation) {
+  if (httpOperationPart) {
+    // body decorated with @multipartBody
+    const httpPart = getHttpPart(context.program, type.type);
+    base.multipartOptions = {
+      isFilePart:
+        isBytesInput ||
+        (httpPart !== undefined && isOrExtendsHttpFile(context.program, httpPart.type)),
+      multi: httpOperationPart.multi,
+      filename: httpOperationPart.filename
+        ? diagnostics.pipe(getSdkModelPropertyType(context, httpOperationPart.filename, operation))
+        : undefined,
+      contentType: httpOperationPart.body.contentTypeProperty
+        ? diagnostics.pipe(
+            getSdkModelPropertyType(context, httpOperationPart.body.contentTypeProperty, operation)
+          )
+        : undefined,
+      defaultContentTypes: httpOperationPart.body.contentTypes,
+    };
+  } else if (operation) {
+    // common body
     const httpOperation = getHttpOperationWithCache(context, operation);
-    operationIsMultipart = Boolean(
+    const operationIsMultipart = Boolean(
       httpOperation && httpOperation.parameters.body?.contentTypes.includes("multipart/form-data")
     );
+    if (operationIsMultipart) {
+      // Currently we only recognize bytes and list of bytes as potential file inputs
+      if (isBytesInput && getEncode(context.program, type)) {
+        diagnostics.add(
+          createDiagnostic({
+            code: "encoding-multipart-bytes",
+            target: type,
+          })
+        );
+      }
+      base.multipartOptions = {
+        isFilePart: isBytesInput,
+        multi: base.type.kind === "array",
+        defaultContentTypes: [],
```

Comment:
Not sure whether TCGC want to provide the `defaultContentTypes`. E.g. nothing for text, `application/octet-stream` for file.

Reply:
Now only JS needs it. If more languages need, TCGC could help calculate it for default values.

Reply:
https://github.com/Azure/typespec-azure/pull/1090#discussion_r1660521295